### ISSUE TITLE
Return a Group of Sym. Already added Makeloss but still get error of 'Need Head Gradient'

### ISSUE THREAD

**Issue by ZijiaLewisLu** [dmlc/mxnet#2486]
Hi, I am building an rnn for segmentation, and I want to pass out c, h and prediction at the same time, but encountered error `Check failed: (info.ref_count) == (0) Because the last operator is not Loss function, head_gradient is required in calling backward`.

This is last part of my code:
    
    .... 
    c = (forget_gate * c) + (in_gate * in_transform)
    h = out_gate * mx.sym.Activation(c, act_type="tanh")

    if dropout > 0.:
        h = mx.sym.Dropout(data=h, p=dropout)

    fc = mx.sym.FullyConnected(
    		data=h, 
    		num_hidden=1*256*256,
            name='pred'
            )

    reshape2 = mx.sym.Reshape(data = fc, target_shape = (0,1,256,256))
    sgmd = mx.sym.Activation(data = reshape2, act_type = 'sigmoid')
    net = mx.sym.Custom(data = sgmd, name = 'softmax', op_type = 'sfmx')
    net = mx.sym.MakeLoss(data = net, name='loss')

    group = mx.sym.Group([net,c,h])
    return group

`sfmx` is an operator I write to deal with the different label in segmentation, and I checked it works normally. I cannot figure out why, with `makeloss`, it still gives me the error. 

Thanks a lot!

**Comment by piiswrong** [comment]
c and h also need head grad. If that's not what you want, use blockgrad on them

**Comment by ZijiaLewisLu** [comment]
Thank you for your quick reply! Issus solved.

**Comment by ZijiaLewisLu** [comment]
Hi, I reopened this issue because I found that blockgrad kill the grad not only for one variable but for all variable after it. I checked , after c and h, all grad become zero. is there a way to solve the issue I mentioned before but also keep the gradient? thanks!

**Comment by sxjscience** [comment]
@ZijiaLewisLu Could you provide some example codes on it so I can debug locally?

**Comment by ZijiaLewisLu** [comment]
@sxjscience  Sure! Thank you a lot!  My code is at ` https://github.com/ZijiaLewisLu/HeartDeep-Kaggle-DSB2`. Please have a look at` /Net/RNN/rnn.py`.  You can commit out all `import ipt`.  The data files are too big so I cannot upload them. For test, you can make up an array of  T(num of time step)*N*1*256*256, and a mark of shape T to tell the net at which time step backprop is required.

**Comment by sxjscience** [comment]
@ZijiaLewisLu I've checked the code. Could you further simplify it and get a minimal example that shows the unexpected behavior of `BlockGrad`?

**Comment by ZijiaLewisLu** [comment]
@sxjscience  Hi, I am trying to simply it, but as I rewrite many functions, it takes a little time... Below is a graph of my original net:
![1](https://cloud.githubusercontent.com/assets/16264284/16355021/4b66029c-3add-11e6-89ab-a2e54c6249f1.jpeg)
 
Then I got the `need head_grad` error. So I added the GradBlock in two places:

![2](https://cloud.githubusercontent.com/assets/16264284/16355028/8c22fd3a-3add-11e6-9e9d-8b8359f63f3a.jpeg)

As shown in the pic, the gradient after the GradBlock becomes zero.  Any suggestions about how can I fix the problem? 

I will upload the simplified net the time I finished it.

**Comment by sxjscience** [comment]
@ZijiaLewisLu I think that the h_{t-1}, c_{t-1} should be directly linked to h_t and c_t. BlockGrad seems to be unnecessary.

**Comment by ZijiaLewisLu** [comment]
@sxjscience  How can I link them together??

**Comment by sxjscience** [comment]
@ZijiaLewisLu You can refer to the LSTM example https://github.com/dmlc/mxnet/blob/master/example/rnn/lstm.py#L17-L40 . Use previous state/cell to generate the current state/cell.

**Comment by ZijiaLewisLu** [comment]
@sxjscience I read the example before. But the state is based on symbol operation. What I hope to do is to add c and h into output so they are in the form of ndarray, and the length of rnn sequence is not needed to be fixed. I am wondering specifically, why I would get the 'need head gradient' error, and is there a way to solve it without GradBlock

**Comment by sxjscience** [comment]
@ZijiaLewisLu Symbol cannot be mixed with NDArray currently. You need to construct the computation logic using Symbolic API and build an executor to do that.
